### ISSUE TITLE
Problem with mutliread of cbor2 object ... (generator pattern).

### ISSUE THREAD

**Issue by ChameleonRed** [agronholm/cbor2#12]
I want to write many objects into one file without use list as proxy since it can be very large amount of data.

How to do it in cbor2 (see cbor works with EOFError pattern)?
It is works in cbor2 looks like bug I get different exception?

```
Traceback (most recent call last):
  File "C:/Users/Cezary Wagner/PycharmProjects/chess-lichess-eval-parse/sandbox/decode_games/s01_pgn_to_cbor.py", line 42, in <module>
    main()
  File "C:/Users/Cezary Wagner/PycharmProjects/chess-lichess-eval-parse/sandbox/decode_games/s01_pgn_to_cbor.py", line 34, in main
    cbor2.load(f)
  File "C:\root\Python36-64\lib\site-packages\cbor2\decoder.py", line 411, in load
    return CBORDecoder(fp, **kwargs).decode()
  File "C:\root\Python36-64\lib\site-packages\cbor2\decoder.py", line 363, in decode
    .format(self.fp.tell(), e))
cbor2.decoder.CBORDecodeError: error reading major type at index 10126175: index out of range
```

```
import datetime

import cbor2
from cbor import cbor

from cw_chess.perfect_pgn import PerfectPgnGamesOffsetsParser


def main():
    start_time = datetime.datetime.utcnow()
    with open('../parser_eval/data/out-6976.txt', newline='') as pgn, \
            open('data/games_offsets.cbor', 'wb') as f:
        pgn_games_offset_parser = PerfectPgnGamesOffsetsParser(pgn)

        for game_offset in pgn_games_offset_parser.games_offsets():
            cbor.dump(game_offset, f)
    end_time = datetime.datetime.utcnow()
    print(end_time - start_time)

    start_time = datetime.datetime.utcnow()
    with open('data/games_offsets.cbor', 'rb') as f:
        while True:
            try:
                cbor.load(f)
            except EOFError:
                break
    end_time = datetime.datetime.utcnow()
    print(end_time - start_time)

    start_time = datetime.datetime.utcnow()
    with open('data/games_offsets.cbor', 'rb') as f:
        while True:
            try:
                cbor2.load(f)
            except EOFError:
                break
    end_time = datetime.datetime.utcnow()
    print(end_time - start_time)


if __name__ == '__main__':
    main()
```

**Comment by agronholm** [comment]
cbor2 does not support streaming. I don't think its README claims that either.

**Comment by ChameleonRed** [comment]
Strange words you talked since it supports streaming I tested code and and it works. Why do you think it not supports?

This code works. Should support.

```
    start_time = datetime.datetime.utcnow()
    games = 0
    with open('data/games_offsets.cbor', 'rb') as f:
        while True:
            try:
                cbor2.load(f)
                games += 1
            except CBORDecodeError:
                break
    end_time = datetime.datetime.utcnow()
    print(end_time - start_time, games)
```

**Comment by ChameleonRed** [comment]
Whatever it should throw EOFErrorinstead CBORDecodeError since it can distinguish.

Other library cbor does not have problem but is little slower and has simpler encoder.

**Comment by agronholm** [comment]
I vaguely recall that this has come up before. I will try to look into this.

On another note, cbor2 has always been slower than cbor in every benchmark I've run. I'm curious how you got these results. I would've thought that cbor's C based codec would give it enough of an advantage over cbor2.

**Comment by agronholm** [comment]
Let me try to understand this. The only reason you would get an EOFError from either library is because you tried to decode a CBOR stream when there wasn't enough data to read. Wouldn't it be easier to check if the stream was at EOF before you start reading a new object?

**Comment by ChameleonRed** [comment]
I am using python3.6 run from PyCharm - serialization was games offsets pairs of (int, int).

For me this results is nothing strange since not know difference but just measure - maybe you are right but I am not tested - speed of code not depend on usage C in most cases :) - weak program in C can be slower than just Python it depend on coder and algorithm.

> I vaguely recall that this has come up before. I will try to look into this.
> 
> On another note, cbor2 has always been slower than cbor in every benchmark I've run. I'm curious how you got these results. I would've thought that cbor's C based codec would give it enough of an advantage over cbor2.

**Comment by agronholm** [comment]
I don't quite understand you. You're saying you measured but didn't test? Then how did you conclude, without measuring, that cbor2 was faster than cbor? Because I **have** measured and it was slower, not faster.

**Comment by agronholm** [comment]
Also, have you tried my approach to deserializing from stream?

**Comment by ChameleonRed** [comment]
For me it is not problem but it is ugly API design and for less skilled coders it is obstacle. Consider that better is API with high usability not speed or functionality what is often not known or forgotten.

My case is processing such files ... - https://database.lichess.org/ - each hash 3G but important is about 10% whatever 10% of 3G = 300M it is huge data do I need do multiple and progressive serialization to keep processing caches instead reading and processing again.

Full processing can take many days do it is important to restart at any step.

I will do such pattern - check if cached, if not check if cached part, if not process part - games offsets, games, processed games, position trees itd.

> Let me try to understand this. The only reason you would get an EOFError from either library is because you tried to decode a CBOR stream when there wasn't enough data to read. Wouldn't it be easier to check if the stream was at EOF before you start reading a new object?

**Comment by agronholm** [comment]
I am hesitant to pass through `EOFException` because that will change the semantics of the decoder and thus require a major version bump.

**Comment by ChameleonRed** [comment]
You can suggest pattern that you can do multiple dumps and multiple load and send some custom exception or pass EOFError - other programmers can learn from documentation that is possible and how to handle it. Now problem is that EOFError != CBORDecodeError.

I think that CBORDecodeError is not signal of EOF but rather high level problem with structure of bytes. That is my problem and this problem is not exists in cbor (they pass EOF and it is clear what i state).

Think about it three days and decide.

> I am hesitant to pass through EOFException because that will change the semantics of the decoder and thus require a major version bump.

**Comment by ChameleonRed** [comment]
I will respond later now I must quit talk (next day maybe).

**Comment by agronholm** [comment]
You could look at the `__cause__` in the exception object if you're using Python 3. That should contain the `EOFException`, I think.

**Comment by ChameleonRed** [comment]
I will try __cause__ since not using it at all. Whatever I still think that some decision takes times if you give some tutorial how to use cbor2 it will better than cbor :)

I think that both libraries are super fast even one will be little slower. The major difference is usability.

**Comment by Sekenre** [comment]
I use cbor2 to compact large time-series measurements stored in CSV
CBOR handles large streams of items as indefinite length arrays. Here's how I use cbor2 to encode them:
```python
@contextmanager
def write_stream(encoder, _type=list):
    """
    Context manager for writing an unknown quantity of records sequentially.

    :param encoder: CBOREncoder object
    :param _type: Type object describing the container, dict types require
        (key, value) arguments, string and byte types require the type to be
        str and bytes respectively, list types can take any serializable value.

    """
    type_map = {list: 0x90, dict:0xb0, str:0x70, bytes: 0x50}
    typecode = type_map.get(_type)
    encoder.write(encode_length(typecode, 0x0f))
    if _type is dict:
        def _encode(key, value):
            encoder.encode(key)
            encoder.encode(value)
    elif _type is list:
        def _encode(value):
            encoder.encode(value)
    else:
        def _encode(value):
            if type(value) is not _type:
                msg = 'Wrong type when writing indefinite {}: {}'
                raise CBOREncodeError(msg.format(_type, type(value)))
            encoder.encode(value)
    yield _encode
    encoder.write(struct.pack('>B', 0xff))
```

And then write to it like this (usually after writing a header):

```python
reader = csv.reader(open('blah.csv', newline=''), delimiter=';')
encoder = CBOREncoder(open('blah.cbor', 'wb'))
encoder.encode({'h':[ <columns>], 'd':';'})
with write_stream(encoder, _type=list) as write:
            for row in reader:
                ts = datetime.strptime(row[0], '%Y-%m-%dT%H:%M:%S').timestamp()
                enc_row = [ts]
                enc_row.extend([ float(x) for x in row[1:] ])
                write(enc_row)
```

Similarly on the decoding end I just read the header, discard the first byte of the indefinite list container and then read complete objects in a stream until reaching the end byte marker:

```python
from cbor2 import CBORDecoder
from cbor2.types import break_marker
    with open('blah.cbor', 'rb') as f:
        decoder = CBORDecoder(f)
        header = decoder.decode()
        f.read(1) # read ahead the single byte denoting an indefinite array
        delimiter = header['d']
        print(delimiter.join(header['h']))
        for items in iter(decoder.decode, break_marker):
            print(delimiter.join(items))
```

**Comment by waveform80** [comment]
A possible compromise on cbor2 raising a different exception might be to use multiple inheritance on the exceptions. I realize this sounds a little odd, but it's a trick I've used a few times in various libraries and might be applicable here...

First a quick re-statement of the major points:

* Most codecs dealing with framed formats (marshal, pickle, cbor) permit reading from a stream repeatedly and simply raise (or pass-through) EOFError when they hit the end of their source.
* cbor2 has similar behaviour, but crucially it raises a different exception: CBORDecodeError, which is specific to the library
* We don't want to change the exception that's raised as users might well be relying upon it.
* At the same time it would be nice if cbor2 shared an API (including the exceptions it raises) with other common codecs.

Now consider if we define a new exception with the type:

```python3
class CBORDecodeEOF(CBORDecodeError, EOFError):
    "Error class raised when decoding unexpectedly reaches EOF."
```

And modify CBORDecoder.read to raise this instead of CBORDecodeError:

```patch
     def read(self, amount):
         """
         Read bytes from the data stream.
 
         :param int amount: the number of bytes to read
         """
         data = self._fp_read(amount)
         if len(data) < amount:
-            raise CBORDecodeError(
+            raise CBORDecodeEOF(
                 'premature end of stream (expected to read {} bytes, got {} '
                 'instead)'.format(amount, len(data)))
 
         return data
```

Existing users can continue catching `CBORDecodeError` because the exception still descends from this. Likewise, those that wish to treat cbor2 like any other Python codec can simply catch `EOFError` because the exception also descends from that.

This is a pattern I've used in a few libraries: define a library-specific root exception type (in this case, that'd be `CBORError` though I usually have mine descend from `Exception`), then create descendants for all the cases required (e.g. `CBORDecodeError`, and `CBORDecodeEOF`) adding whatever Python exception class I'd *normally* expect to that specific error's ancestor list (commonly this is `ValueError` or `AttributeError`).

The result is that users can either opt to catch all errors related to my specific library (by catching the library-specific root), or the common Python exception they'd expect (`EOFError` in this case), or, if required, the precise exception defined in the circumstance (`CBORDecodeEOF`).

I'm happy to put together a quick PR for this, if it's agreeable? I'm about half-way thru reviewing the docs at the mo (there's a few minor changes needed to fix a few examples; hope to have a PR finished for tomorrow night).
